cfbs/analyze.py

```diff
@@ -0,0 +1,597 @@
+)
+from cfbs.utils import (
+    FetchError,
+    cfbs_dir,
+    fetch_url,
+    file_sha256,
+    get_json,
+    get_or_read_json,
+    immediate_subdirectories,
+    mkdir,
+    user_error,
+)
+
+
+def path_components(path):
+    """Returns a list of path components of `path`.
+
+    The first component is `""` for a relative path starting with a separator. On Windows, if `path` begins with n backslashes, the first n components will be `""`.
+
+    The last component is the filename, trailing separators do not affect the result."""
+    norm_path = os.path.normpath(path)
+
+    dir_components = norm_path.split(os.sep)
+
+    return dir_components
+
+
+def is_path_component(path, component):
+    """Returns whether `component` is a path component of `path`."""
+    p_components = path_components(path)
+
+    # check if `component` is a directory or a file
+    if component[-1] == "/":
+        # strip the suffixed directory slash
+        component = component[:-1]
+
+        return component in p_components[:-1]
+    else:
+        return component == p_components[-1]
+
+
+def contains_ignored_components(path, ignored_components):
+    """Returns whether `path` contains any of the path components in `ignored_components`."""
+    for i_comp in ignored_components:
+        if is_path_component(path, i_comp):
+            return True
+
+    return False
+
+
+DEFAULT_CHECKSUMS_DICT = {"checksums": {}}
+DEFAULT_FILES_DICT = {"files": {}}
+
+
+def checksums_files(
+    files_dir_path,
+    checksums_dict=None,
+    files_dict=None,
+    ignored_path_components=[],
+):
+    if checksums_dict is None:
+        checksums_dict = DEFAULT_CHECKSUMS_DICT
+    if files_dict is None:
+        files_dict = DEFAULT_FILES_DICT
+
+    for root, _, files in os.walk(files_dir_path):
+        for name in files:
+            full_relpath = os.path.join(root, name)
+            tarball_relpath = os.path.relpath(full_relpath, files_dir_path)
+            file_checksum = file_sha256(full_relpath)
+
+            if contains_ignored_components(full_relpath, ignored_path_components):
+                continue
+
+            if not file_checksum in checksums_dict["checksums"]:
+                checksums_dict["checksums"][file_checksum] = set()
+            checksums_dict["checksums"][file_checksum].add(tarball_relpath)
+
+            if not tarball_relpath in files_dict["files"]:
+                files_dict["files"][tarball_relpath] = set()
+            files_dict["files"][tarball_relpath].add(file_checksum)
+
+    return checksums_dict, files_dict
+
+
+def mpf_vcf_dicts(offline=False):
+    REPO_OWNER = "cfengine"
+    REPO_NAME = "release-information"
+
+    REPO_OWNERNAME = REPO_OWNER + "/" + REPO_NAME
+    RI_SUBDIRS = "downloads/github.com/" + REPO_OWNERNAME + "/archive/refs/tags/"
+
+    if offline:
+        ERROR_MESSAGE = "MPF release information not found. Provide the release information, for example by running 'cfbs analyze' without '--offline'."
```

Comment:
In a message to the user I wonder if MPF should be Masterfiles or Masterfiles Policy Framework.